R/sim_network_bp.R

```diff
@@ -0,0 +1,126 @@
+#' Simulate a random network branching process model with a probability of
+#' infection for each contact
+#'
+#' @description
+#' Simulate a branching process on a infinite network where the contact
+#' distribution provides a function to sample the number of contacts of each
+#' individual in the simulation. Each contact is then infected with the
+#' probability of infection. The time between each infection is determined
+#' by the contact interval function.
+#'
+#' @details
+#' The contact distribution sampled takes the network effect
+#' \eqn{q(n) \sim (n + 1)p(n + 1)} where \eqn{p(n)} is the probability
+#' density function of a distribution, e.g., Poisson or Negative binomial.
+#' That is to say, the probability of having choosing a contact at random by
+#' following up a contact chooses individuals with a probability proportional
+#' to their number of contacts. The plus one is because one of the contacts
+#' was "used" to infect the person.
+#'
+#' @inheritParams sim_linelist
+#'
+#' @return A `<data.frame>` with the contact and transmission chain data.
+#' @keywords internal
+.sim_network_bp <- function(contact_distribution,
+                            contact_interval,
+                            prob_infect,
+                            add_names) {
+
+  # initialise data object
+  ancestor <- vector(mode = "integer", 1e5)
+  generation <- vector(mode = "integer", 1e5)
+  infected <- vector(mode = "integer", 1e5)
+  time <- vector(mode = "double", 1e5)
+
+  # store initial individual
+  ancestor[1] <- NA_integer_
+  generation[1] <- 1L
+  # 1 is infected, 0 is non-infected contact
+  infected[1] <- 1L
+  time[1] <- 0
+
+  # initialise counters
+  next_gen_size <- 1L
+  chain_generation <- 1L
+  chain_size <- 1L
+  chain_length <- 1L
+  ancestor_idx <- 1L
+  prev_ancestors <- 1L
+
+  # run loop until no more individuals are sampled
+  while (next_gen_size > 0) {
+    # sample contact distribution (excess degree distribution)
+    q <- contact_distribution(0:100 + 1) * (0:100 + 1)
+    q <- q / sum(q)
+    contacts <- sample(0:100, size = next_gen_size, replace = TRUE, prob = q)
+
+    # add contacts if sampled
+    if (sum(contacts) > 0L) {
+      chain_size <- chain_size + sum(contacts)
+      chain_length <- chain_length + any(contacts >= 1L)
+      chain_generation <- chain_generation + 1L
+
+      if (add_names && chain_size > 5180) {
```

Comment:
I'm looking into a possible alternative for {randomNames} as I'm not happy with the current limitations of number of line list rows. If there are no possible replacements I will try with a `try()` or `purrr::safely()`. I agree with your point that they allow the condition for when something errors to change without having to update the code, but they would pass an uninformative error from {randomNames} to the user, and I think the benefit of setting the condition ourselves has the benefit that we can return an informative warning and return the data early rather than erroring.

If I'm missing something from the `try()` or `purrr::safely()` (I haven't used this function before) approach please let me know.

Reply:
Thankfully this `if` statement with a `warning()` and `break` has been removed since: 69ff03f4a0d8b64689201b8a8ce30cb22267746c. A work around resulting from https://github.com/CenterForAssessment/randomNames/issues/83 has meant the number of unique names that can be produced is much higher (several million).